### PR TITLE
test(storage): add support for grpc

### PR DESCRIPTION
**Pull request by vnvo2409**
This PR:
- adds support for grpc.
- makes `grpc_integration_test` testing with both `media` and `metadata` since it is our ultimate goal.
- `CI` changes

`gevent` and `gRPC` can not work together, so I change the worker to `sync` and use `threads`. I also add a endpoint `start_grpc` to start the grpc server. The reason is `Gunicorn` will spawn a new subprocess when running `Flask` server. If we start `gRPC` server before the spawn of the subprocess, it's nearly impossible to share the `database` with the new subprocess because Python will copy everything in the memory from the parent process to the subprocess ( So we have 2 separate instance of `database` ). The endpoint will start the `gRPC` server in the same subprocess so there is only one instance of `database`.

Part of #4751 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414)

### CODE REVIEW

**Comment by vnvo2409** [comment]
@coryan 
Could you kick-off a CI please ?

**Comment by vnvo2409** [comment]
Both `bazel` and `cmake` passed ( I forgot to remove the `storage_grpc_samples` out of the tests against production in `cmake` but it passed with `bazel` ).

`clang-tidy` failed because of formating.

What do you think about this PR ? @coryan

**Review by coryan — COMMENTED** [review]
> What do you think about this PR ?

I am sorry I have been less responsive than usual, I am super busy with other stuff. The PR is looking good, please fix formatting and there are some questions below.  Let me know those make no sense. 

> I have to disable one test because it always runs against production:

SGTM

> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-:-MLO_quo8-AZ6RDkRynW:bj3lj42)* status: all files reviewed, 4 unresolved discussions (waiting on @vnvo2409)

___
*[google/cloud/storage/emulator/emulator.py, line 45 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-MLOPT8hEj1VI8DxHEQa:-MLOPT8hEj1VI8DxHEQb:b-pky5vb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b446c9eaf952e81a6847a9ce7b71924439811fd6/google/cloud/storage/emulator/emulator.py#L45)):*
> ```Python
> 
> @root.route("/start_grpc")
> def start_grpc():
> ```

Is there a reason we cannot start this immediately?  If there is, can you add a comment explaining why?
___
*[google/cloud/storage/emulator/grpc_server.py, line 30 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-MLOPbLP4KNry7_V5tF8:-MLOPbLP4KNry7_V5tF9:b5vngxr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b446c9eaf952e81a6847a9ce7b71924439811fd6/google/cloud/storage/emulator/grpc_server.py#L30)):*
> ```Python
> 
> 
> class StorageServicer(storage_pb2_grpc.StorageServicer):
> ```

FYI... "Servicer" is a weird name, I see that it is the convention in Python...  :man_shrugging: 
___
*[google/cloud/storage/emulator/grpc_server.py, line 159 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-MLOPnGL7AvhYYfNSenW:-MLOPnGL7AvhYYfNSenX:b-7p4vua) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b446c9eaf952e81a6847a9ce7b71924439811fd6/google/cloud/storage/emulator/grpc_server.py#L159)):*
> ```Python
>     db = database
>     storage_pb2_grpc.add_StorageServicer_to_server(StorageServicer(), server)
>     server.add_insecure_port("[::]:" + port)
> ```

I think this is assuming IPv6, is there a reason why `"localhost:" + port` does not work?
___
*[google/cloud/storage/emulator/gcs/holder.py, line 94 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-MLO_RvP6u1uEdW70kaa:-MLO_RvP6u1uEdW70kab:b6q17jx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b446c9eaf952e81a6847a9ce7b71924439811fd6/google/cloud/storage/emulator/gcs/holder.py#L94)):*
> ```Python
>             ("%s/o/%s" % (bucket.name, metadata.name)).encode("utf-8")
>         ).hexdigest()
>         fake_request = utils.common.FakeRequest.init_protobuf(request, context)
> ```

Could you add a comment explaining  what is going on here?
___
*[google/cloud/storage/tools/run_testbench_utils.sh, line 109 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-MLO_E2e5PQzKaA-xxDl:-MLO_E2f1Sx9b2jOIjJ5:b-golple) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b446c9eaf952e81a6847a9ce7b71924439811fd6/google/cloud/storage/tools/run_testbench_utils.sh#L109)):*
> ```Shell
>   fi
> 
>   grpc_port=8000
> ```

Could we make this an argument received by the function?  And could we make it accept `0` and pick a random port for us?

**Comment by vnvo2409** [comment]
> Is there a reason we cannot start this immediately? If there is, can you add a comment explaining why?

Changed to `localhost`. The homepage uses IPv6 so I copied it.

---

Sum up what I pushed:
f0b8c8f: Formatting

---

964c83e: 
I accidentally pass `None` to `context` when working with gRPC ( probaly because of copy-paste ) so this commit fix it. In addition, It adds a feature to update the `FakeRequest` based on scalar fields in the protobuf request.
```python
            ("%s/o/%s" % (bucket.name, metadata.name)).encode("utf-8")
        ).hexdigest()
        fake_request = utils.common.FakeRequest.init_protobuf(request, context)
```
> Could you add a comment explaining what is going on here?

`init_protobuf` is used to flatten protobuf request. For an example, `InsertObjectRequest`'s structure is
```protobuf
message InsertObjectRequest {
  // The first message of each stream should set one of the following.
  oneof first_message {
    // For resumable uploads. This should be the `upload_id` returned from a
    // call to `StartResumableWriteResponse`.
    string upload_id = 1;

    // For non-resumable uploads. Describes the overall upload, including the
    // destination bucket and object name, preconditions, etc.
    InsertObjectSpec insert_object_spec = 2;
  }
  CommonObjectRequestParams common_object_request_params = 3;

  // A set of parameters common to all Storage API requests.
  CommonRequestParams common_request_params = 4;
  // ...
}
message InsertObjectSpec {
  // Destination object, including its name and its metadata.
  Object resource = 1;

  // Apply a predefined set of access controls to this object.
  CommonEnums.PredefinedObjectAcl predefined_acl = 2;
  // ....
  CommonEnums.Projection projection = 7;
}

```
If we want to extract `predefined_acl`, we will have to do `request-> insert_object_spec-> predefined_acl` which will eventually break all the utils functions. So the `init_protobuf` will create a `FakeRequest` whose structure is something like:
```protobuf
{
//...
CommonEnums.PredefinedObjectAcl predefined_acl = 2;
CommonObjectRequestParams common_object_request_params = 3;
CommonRequestParams common_request_params = 4;
//...
}
```
We could save that `fake_request` and pass it around without worrying about the nested message.

 ---

61d42c8
> Could we make this an argument received by the function? And could we make it accept 0 and pick a random port for us?

Done but I am not so comfortable with `bash` so we will have to wait for the CI.

**Comment by vnvo2409** [comment]
```python
@root.route("/start_grpc")
def start_grpc():
```
> Is there a reason we cannot start this immediately? If there is, can you add a comment explaining why?

The reason is `Gunicorn` will spawn a new subprocess ( a worker ) when running Flask server. If we start `gRPC` server before the spawn of the subprocess, it's nearly impossible to share the `database` with the new subprocess because Python will copy everything in the memory from the parent process to the subprocess ( So we have 2 separate instance of `database` ). The endpoint will start the `gRPC` server in the same subprocess so there is only one instance of `database`. Please let me know if there is something that is unclear.

**Review by coryan — CHANGES_REQUESTED** [review]
The changes look great, thank you.  Just this one thing:

>> Is there a reason we cannot start this immediately? If there is, can you add a comment explaining why?
> 
> The reason is Gunicorn will spawn a new subprocess ( a worker ) when running Flask server. If we start gRPC server before the spawn of the subprocess, it's nearly impossible to share the database with the new subprocess because Python will copy everything in the memory from the parent process to the subprocess ( So we have 2 separate instance of database ). The endpoint will start the gRPC server in the same subprocess so there is only one instance of database. Please let me know if there is something that is unclear.

Your explanation makes perfect sense. But our future-selves will have forgotten it. Putting that comment in the code will make it easier to remember why it is there.

> Reviewed 5 of 6 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-:-MLSkkb8BTZHwXIdtoVJ:b-jo7fr9)* status: all files reviewed, 3 unresolved discussions (waiting on @vnvo2409)

___
*[google/cloud/storage/tools/run_testbench_utils.sh, line 111 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-MLSlFeI6ozAo33RcMZb:-MLSlFeI6ozAo33RcMZc:b-ibhdji) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4e83ba689a0d384c670df997da9b57fae9aea036/google/cloud/storage/tools/run_testbench_utils.sh#L111)):*
> ```Shell
>   port="${2:-0}"
>   local grpc_port=""
>   grpc_port=$(curl -s "http://localhost:${testbench_port}/start_grpc?port=${port}")
> ```

I wonder if we need a retry loop here?  I think maybe you can add `--retry-times=5` for curl to do the retry for you.

**Comment by vnvo2409** [comment]
@coryan 
Is there a way to get the `testbench.log` inside kokoro :/ `kokoro/docker/coverage` failed and it only shows `Internal Server Error`. I can not reproduce it locally.

**Comment by coryan** [comment]
> @coryan
> Is there a way to get the `testbench.log` inside kokoro

Unfortunately not. These are all discarded when the build finishes.

> In addition, Issues like #5284 could be solved easier if we have access to `testbench.log`.

Ack.

>If we could not download the log, how about `cat testbench.log` before stopping kokoro if the integration tests failed ?

That SGTM, feel free to add this to your PR so we can troubleshoot.  In case it  there is a bash function to dump only the first N and last N lines of a lot already:

https://github.com/googleapis/google-cloud-cpp/blob/master/ci/define-dump-log.sh

File: google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh
Comment (vnvo2409) [inline google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh:95]: Is it the right location of `testbench.log` ?

Comment (coryan) [reply google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh:95]: Ooops, no.  I think it is placed in `$HOME`, see above just before `start_testbench`.

Comment (vnvo2409) [reply google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh:95]: Could I fix it in the next PR ?

Comment (coryan) [reply google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh:95]: Yes, absolutely, I thought you needed it to troubleshoot something.

Comment (vnvo2409) [reply google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh:95]: This time, no test failed so I cannot troubleshoot anything. Just to be sure, the extract location of `testbench.log` is `$HOME/testbench.log` right ?

**Comment by coryan** [comment]
___
*[google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh, line 95 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-MLNlXRlGr7Wkmn8tX_Y-r4-95:-MLTV9zZ5ahFRguDSY06:b-903fw7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c20dae017e8950e9a214989984b846ab237d4390/google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh#L95)):*
<details><summary><i>Previously, vnvo2409 (Vo Van Nghia) wrote…</i></summary><blockquote>

This time, no test failed so I cannot troubleshoot anything. Just to be sure, the extract location of `testbench.log` is `$HOME/testbench.log` right ?
</blockquote></details>

I believe that is correct.

**Review by coryan — COMMENTED** [review]
I created #5428 to track restoring this test to the code coverage build:

>            TEST_F(GrpcIntegrationTest, GetObjectMediaNotFound) { 

Otherwise, this :lgtm_strong: Thanks so much for making this multi-step contribution.

> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5414#-:-MLTUtgJ33bxNIDWYI3V:b-2zx18n)* status: all files reviewed, 2 unresolved discussions (waiting on @vnvo2409)

**Comment by coryan** [comment]
I am reverting this PR, sorry, the builds are failing after the merge.

**Pull request merged**
